src/ErrorBoundary.tsx

```diff
@@ -1,89 +1,49 @@
-  };
+type GenericProps = Record<string, unknown>;
 
-  public componentDidUpdate(prevProps: Readonly<P & ErrorBoundaryProps<P>>) {
-    if (
-      this.state.error !== null &&
-      this.props.recoverableProps !== void 0 &&
-      Array.isArray(this.props.recoverableProps)
-    ) {
-      for (const recoverableProp of this.props.recoverableProps) {
-        if (prevProps[recoverableProp] !== this.props[recoverableProp]) {
-          this.setError(null);
-          break;
-        }
-      }
-    }
-  }
-
-  public componentDidCatch(error: Error, errorInfo: React.ErrorInfo) {
-    this.setError(error, errorInfo.componentStack);
-    this.handleError(error, errorInfo);
-  }
-
-  protected handleError(error: Error, errorInfo: React.ErrorInfo | null) {
-    if (this.props.reportErrors !== false) {
-      try {
-        if (errorInfo !== null) {
-          this.context.reporter.error(error.message, { errorInfo });
-        } else {
-          this.context.reporter.error(error);
-        }
-      } catch (ex) {
-        console.error('Error could not be reported', ex);
-      }
-    }
-
-    if (this.props.onError !== void 0) {
-      try {
-        this.props.onError(error, errorInfo && errorInfo.componentStack);
-      } catch {
-        // happens
-      }
-    }
-  }
+function usePrev(value: GenericProps) {
+  const ref = React.useRef<GenericProps>();
+  React.useEffect(() => {
+    ref.current = value;
+  }, [value]);
 
-  public throwError = (error: Error) => {
-    this.setError(error);
-    this.handleError(error, null);
-  };
+  return ref.current;
+}
 
-  protected setError(error: Error | null, componentStack: string | null = null) {
-    this.setState({ error, componentStack });
-  }
+export const ErrorBoundary: React.FC<ErrorBoundaryProps<GenericProps> & GenericProps> = props => {
+  const context = React.useContext(ErrorBoundaryContext);
+  const fallback = props.FallbackComponent || context.FallbackComponent || FallbackComponent;
+  const ActualFallback = React.useMemo<FallbackRender>(() => wrapFallback(fallback), [fallback]);
 
-  protected recover = () => {
-    if (this.state.error !== null) {
-      this.setError(null);
-    } else if (process.env.NODE_ENV !== 'production') {
-      console.warn('Component has not crashed. Recovering is a no-op in such case');
-    }
-  };
+  const boundaryRef = React.useRef<SentryErrorBoundary | null>(null);
+  const prevProps = usePrev(props);
 
-  public render() {
-    const {
-      props: { FallbackComponent: Fallback = this.context.FallbackComponent || FallbackComponent, children },
-      state: { error, componentStack },
-    } = this;
+  React.useEffect(() => {
+    const boundary = boundaryRef.current;
+    if (!boundary || !prevProps || !boundary.state.error || !props.recoverableProps) return;
 
-    if (error !== null) {
-      return <Fallback error={error} componentStack={componentStack} tryRecovering={this.recover} />;
+    for (const recoverableProp of props.recoverableProps) {
```

Comment:
@marbemac @lottamus whatcha think about dropping `recoverableProps`?
I know it's used here and there, but one can use the `reload` button in the fallback if needed.

This seems to work reasonably though, so can keep it.

Reply:
How "in use" is `recoverableProps` in our ecosystem? What are the benefit(s) of it versus just leveraging the out of the box `fallback` + `resetError` and `onReset` functionality that sentry error boundary exposes out of the box?

Reply:
Ah no actually this is a pretty good feature IMHO - would be nice to keep if it doesn't introduce much complexity etc.

Reply:
It doesn't - and it seems to be working correctly, verified it in json-schema-viewer & the storybook here in react-error-boundary.